### ISSUE TITLE
Should SSL_set0_verify_cert_store() be removed from the API?

### ISSUE THREAD

**Issue by ywkaras** [openssl/openssl#12716]
SSL_free() calls ssl_cert_free(), which calls X509_STORE_free() on the store of verify certs, set by this API call.  I did not realize this, so spent several hours figuring out why I got a segment violation on the second handshake using a particular store.  When I switched to the set1 version, worked fine.  I can't think of a scenario where the set0 function would be useful.

Now that I'm using the set1 function, what if the handshake doesn't get far enough so that the client sends a cert?  Will ssl_cert_free() not be called in that case, potentially causing the X509 store to eventually leak?

Minimally, I would suggest adding a warning to https://www.openssl.org/docs/manmaster/man3/SSL_set0_verify_cert_store.html that SSL_free() will (may?) cause a call to X509_STORE_free() on the provided store.

**Comment by mattcaswell** [comment]
Sounds like a documentation issue to me. The set0 vs set1 semantics is common place throughout OpenSSL - but probably is confusing to users not familiar with the idiom. The documentation should explicitly state the memory management impacts of these calls.

**Comment by ywkaras** [comment]
@mattcaswell can you briefly describe a scenario where you would want to use the set0 version?  I assume that would mean that SSL_free would never be called on the SSL object whose client verify store is being set.

**Comment by ywkaras** [comment]
OK I think I get it now.  set0 and set1 both make the SSL object a shared owner of the store.  The difference is that set0 transfers an owning reference into the SSL object, and set1 puts a new owning reference into the SSL object.  Would be helpful if the documentation clarified this.  Doc for SSL_free could make it explicit it calls the appropriate free function to remove the SSL objects owning references.

**Comment by mattcaswell** [comment]
Agreed - the documentation could be improved. Suggested improvements in a PR welcome!

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
